observatory_platform/telescopes/crossref_events.py

```diff
@@ -0,0 +1,701 @@
+        need to be obtained)
+        """
+        self.start_date = start_date
+        self.end_date = end_date
+        self.first_release = first_release
+
+        if CrossrefEventsTelescope.DOWNLOAD_MODE == 'parallel':
+            batches = self.batch_dates
+        elif CrossrefEventsTelescope.DOWNLOAD_MODE == 'sequential':
+            batches = [(start_date, end_date)]
+        else:
+            raise AirflowException(f'Download mode has to be either "sequential" or "parallel", '
+                                   f'not "{CrossrefEventsTelescope.DOWNLOAD_MODE}"')
+
+        self.urls = []
+        edited_deleted = False
+        for batch in batches:
+            start_date = batch[0]
+            end_date = batch[1]
+            event_type_urls = [CrossrefEventsTelescope.EVENTS_URL.format(start_date=start_date, end_date=end_date)]
+            # only add edited/deleted urls to first batch, since no end date can be specified
+            if edited_deleted or first_release:
+                pass
+            else:
+                event_type_urls.append(CrossrefEventsTelescope.EDITED_URL.format(start_date=start_date))
+                event_type_urls.append(CrossrefEventsTelescope.DELETED_URL.format(start_date=start_date))
+                edited_deleted = True
+            self.urls.append(event_type_urls)
+
+    @property
+    def batch_dates(self) -> list:
+        """
+        Create batches of time periods, based on the start and end date of the release and the max number of
+        processes available.
+
+        :return: List of batches, where each batch is a tuple of (start_date, end_date). Both dates are strings in
+        format YYYY-MM-DD
+        """
+        start_date = self.start_date.date()
+        end_date = self.end_date.date()
+        max_processes = CrossrefEventsTelescope.MAX_PROCESSES
+
+        # number of days between start and end, add 1, because end date is included
+        total_no_days = (end_date - start_date).days + 1
+
+        # number of days in each batch, rounded down, minimal is 1
+        batch_no_days = max(1, round(total_no_days / max_processes))
+
+        batches = []
+        batch_start_date = start_date
+        no_batches = total_no_days if total_no_days < max_processes else max_processes
+        # subtract one for possible final batch
+        for x in range(no_batches - 1):
+            batch_end_date = batch_start_date + timedelta(days=batch_no_days - 1)
+            batches.append((batch_start_date.strftime("%Y-%m-%d"), batch_end_date.strftime("%Y-%m-%d")))
+            # end date is included, so start date should be one day later
+            batch_start_date = batch_end_date + timedelta(days=1)
+
+        # check if final batch has to be added
+        if batch_start_date <= end_date:
+            batches.append((batch_start_date.strftime("%Y-%m-%d"), end_date.strftime("%Y-%m-%d")))
+        return batches
```

Comment:
It might be simpler to use the Pendulum range function since we are already using Pendulum: https://pendulum.eustace.io/docs/#range

For example (from the Pendulum documentation): 
```python
import pendulum

start = pendulum.datetime(2000, 1, 1)
end = pendulum.datetime(2000, 1, 10)

period = pendulum.period(start, end)

for dt in period.range('days'):
    print(dt)
```

Reply:
I've used the pendulum range functionality and replaced it with this. I'm not sure if it is much simpler, might be a bit more readable.

```python
batches = []
period = pendulum.period(start_date, end_date)
for i, dt in enumerate(period.range('days', batch_no_days)):
    batch_start_date = dt.strftime("%Y-%m-%d")
    # if final batch or end date is reached before end of period, use end date of release instead
    if i == (max_processes - 1) or dt == end_date:
        batch_end_date = end_date.strftime("%Y-%m-%d")
        batches.append((batch_start_date, batch_end_date))
        break
    # use end date of period
    else:
        # end date is included, so subtract 1 day from batch_no_days
        batch_end_date = (dt + timedelta(days=batch_no_days - 1)).strftime("%Y-%m-%d")
        batches.append((batch_start_date, batch_end_date
```